Gyro does not suspend creation of a resource when there is a dependency

**Describe the bug**
Gyro does not suspend the creation of azure::network sub-resources when they're dependent on another resource. The following is an failed gyro up:

```
 ezapata@admins-MacBook-Pro  ~/Documents/projects/ops-test/ops/gyro-azure   master ●  gyro up
↓ Loading plugin: gyro:gyro-azure-provider:0.99.1-SNAPSHOT
↓ Loading plugin: gyro:gyro-ssh-plugin:0.99.1-SNAPSHOT
⟳ Refreshed resources: 0

Looking for changes...

+ Create azure::dns-zone public-zone
+ Create azure::resource-group resource-group (ops-test-v1)
+ Create azure::network network
    + Create subnet private
    + Create subnet public
    + Create subnet lower
+ Create azure::route-table public
    + Create route internet
+ Create azure::route-table private
    + Create route nat
+ Create azure::route-table lower
    + Create route internet

Are you sure you want to change resources? (y/N) y

+ Creating azure::dns-zone public-zone OK
+ Creating azure::resource-group resource-group (ops-test-v1) OK
+ Creating azure::network network OK
+ Creating subnet private for azure::network network
Error: Can't create subnet private resource!
```

**To Reproduce**
Run a gyro up with the following configs.
network.gyro:
```
azure::network network
    name: "Network-test"
    resource-group: $(azure::resource-group resource-group)
    address-spaces:  [
        "10.0.0.0/16"
    ]
    subnet
        address-prefix: 10.0.0.0/26
        name: "private"
        route-table: $(azure::route-table private)
    end
end
```
routes.gyro:
```
azure::route-table private
    bgp-route-propagation-disabled: true
    resource-group: $(azure::resource-group resource-group)

    route
        destination-address-prefix: "0.0.0.0/0"
        name: "nat"
        next-hop-type: "VirtualAppliance"
        next-hop-ip-address: '10.0.0.10'
    end
end
```

**Expected behavior**
The creation of the network resource should be done after the route-table has been created.

**Stacktrace**
```
Caused by: Unexpected error: com.microsoft.azure.CloudException: Cannot parse the request.: Cannot parse the request.
	at com.microsoft.azure.AzureClient.createExceptionFromResponse(AzureClient.java:740)
	at com.microsoft.azure.AzureClient.access$100(AzureClient.java:33)
	at com.microsoft.azure.AzureClient$3.call(AzureClient.java:160)
	at com.microsoft.azure.AzureClient$3.call(AzureClient.java:157)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69)
	at retrofit2.adapter.rxjava.CallArbiter.deliverResponse(CallArbiter.java:120)
	at retrofit2.adapter.rxjava.CallArbiter.emitResponse(CallArbiter.java:102)
	at retrofit2.adapter.rxjava.CallExecuteOnSubscribe.call(CallExecuteOnSubscribe.java:46)
	at retrofit2.adapter.rxjava.CallExecuteOnSubscribe.call(CallExecuteOnSubscribe.java:24)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeSingle.call(OnSubscribeSingle.java:81)
	at rx.internal.operators.OnSubscribeSingle.call(OnSubscribeSingle.java:27)
	at rx.internal.operators.SingleToObservable.call(SingleToObservable.java:39)
	at rx.internal.operators.SingleToObservable.call(SingleToObservable.java:27)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.DeferredScalarSubscriber.subscribeTo(DeferredScalarSubscriber.java:153)
	at rx.internal.operators.OnSubscribeTakeLastOne.call(OnSubscribeTakeLastOne.java:32)
	at rx.internal.operators.OnSubscribeTakeLastOne.call(OnSubscribeTakeLastOne.java:22)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OperatorSubscribeOn$SubscribeOnSubscriber.call(OperatorSubscribeOn.java:100)
	at rx.internal.schedulers.CachedThreadScheduler$EventLoopWorker$1.call(CachedThreadScheduler.java:230)
	at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: rx.exceptions.OnErrorThrowable$OnNextValue: OnError while emitting onNext value: retrofit2.Response.class
	at rx.exceptions.OnErrorThrowable.addValueAsLastCause(OnErrorThrowable.java:118)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:73)
	... 47 more
```

--- a/core/src/main/java/gyro/core/command/UpCommand.java
+++ b/core/src/main/java/gyro/core/command/UpCommand.java
@@ -38,8 +38,8 @@ public class UpCommand extends AbstractConfigCommand {
 
         while (true) {
             Diff diff = new Diff(
-                current.findResourcesIn(current.getLoadFiles()),
-                pending.findResourcesIn(pending.getLoadFiles()));
+                current.findSortedResourcesIn(current.getLoadFiles()),
+                pending.findSortedResourcesIn(pending.getLoadFiles()));
 
             diff.diff();
 

--- a/core/src/main/java/gyro/core/scope/FileScope.java
+++ b/core/src/main/java/gyro/core/scope/FileScope.java
@@ -16,7 +16,15 @@
 
 package gyro.core.scope;
 
+import java.util.AbstractCollection;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
 import com.google.common.base.Preconditions;
+import gyro.core.resource.Resource;
 
 public class FileScope extends Scope {
 
@@ -32,4 +40,106 @@ public class FileScope extends Scope {
         return file;
     }
 
+    @Override
+    public void clear() {
+        if (!(this instanceof RootScope) && getRootScope() != null) {
+            keySet().forEach(k -> getRootScope().getResources().remove(k));
+        }
+
+        super.clear();
+    }
+
+    @Override
+    public Set<Entry<String, Object>> entrySet() {
+        return new AbstractSet<Entry<String, Object>>() {
+
+            @Override
+            public void clear() {
+                FileScope.this.clear();
+            }
+
+            @Override
+            public Iterator<Map.Entry<String, Object>> iterator() {
+                return FileScope.super.entrySet().iterator();
+            }
+
+            @Override
+            public int size() {
+                return FileScope.super.entrySet().size();
+            }
+        };
+    }
+
+    @Override
+    public Set<String> keySet() {
+        return new AbstractSet<String>() {
+
+            @Override
+            public void clear() {
+                FileScope.this.clear();
+            }
+
+            @Override
+            public Iterator<String> iterator() {
+                return FileScope.super.keySet().iterator();
+            }
+
+            @Override
+            public int size() {
+                return FileScope.super.keySet().size();
+            }
+        };
+    }
+
+    @Override
+    public Object put(String key, Object value) {
+        if (!(this instanceof RootScope) && getRootScope() != null && value instanceof Resource) {
+            getRootScope().getResources().put(key, (Resource) value);
+        }
+
+        return super.put(key, value);
+    }
+
+    @Override
+    public void putAll(Map<? extends String, ?> other) {
+        if (!(this instanceof RootScope) && getRootScope() != null) {
+            other.forEach((key, value) -> {
+                if (value instanceof Resource) {
+                    getRootScope().getResources().put(key, (Resource) value);
+                }
+            });
+        }
+
+        super.putAll(other);
+    }
+
+    @Override
+    public Object remove(Object key) {
+        if (!(this instanceof RootScope) && getRootScope() != null) {
+            getRootScope().getResources().remove(key);
+        }
+
+        return super.remove(key);
+    }
+
+    @Override
+    public Collection<Object> values() {
+        return new AbstractCollection<Object>() {
+
+            @Override
+            public void clear() {
+                FileScope.this.clear();
+            }
+
+            @Override
+            public Iterator<Object> iterator() {
+                return FileScope.super.values().iterator();
+            }
+
+            @Override
+            public int size() {
+                return FileScope.super.values().size();
+            }
+        };
+    }
 }

--- a/core/src/main/java/gyro/core/scope/NodeEvaluator.java
+++ b/core/src/main/java/gyro/core/scope/NodeEvaluator.java
@@ -525,6 +525,7 @@ public class NodeEvaluator implements NodeVisitor<Scope, Object, RuntimeExceptio
             bodyScope.process(resource);
             file.put(fullName, resource);
             file.putLocation(fullName, node);
+            file.getRootScope().getResources().add(resource);
 
         } else if (value instanceof ResourceVisitor) {
             ((ResourceVisitor) value).visit(name, bodyScope);

--- a/core/src/main/java/gyro/core/scope/NodeEvaluator.java
+++ b/core/src/main/java/gyro/core/scope/NodeEvaluator.java
@@ -525,7 +525,6 @@ public class NodeEvaluator implements NodeVisitor<Scope, Object, RuntimeExceptio
             bodyScope.process(resource);
             file.put(fullName, resource);
             file.putLocation(fullName, node);
-            file.getRootScope().getResources().add(resource);
 
         } else if (value instanceof ResourceVisitor) {
             ((ResourceVisitor) value).visit(name, bodyScope);
@@ -749,7 +748,7 @@ public class NodeEvaluator implements NodeVisitor<Scope, Object, RuntimeExceptio
                 throw new WildcardDefer(node, referenceName);
             }
 
-            Stream<Resource> s = root.findResources()
+            Stream<Resource> s = root.findSortedResources()
                 .stream()
                 .filter(r -> referenceName.equals(DiffableType.getInstance(r.getClass()).getName()));
 

--- a/core/src/main/java/gyro/core/scope/RootScope.java
+++ b/core/src/main/java/gyro/core/scope/RootScope.java
@@ -102,6 +102,7 @@ public class RootScope extends FileScope {
     private final FileBackend backend;
     private final RootScope current;
     private final Set<String> loadFiles;
+    private final Set<Resource> resources = new LinkedHashSet<>();
     private final List<FileScope> fileScopes = new ArrayList<>();
 
     public RootScope(String file, FileBackend backend, RootScope current, Set<String> loadFiles) {
@@ -195,6 +196,10 @@ public class RootScope extends FileScope {
         return loadFiles;
     }
 
+    public Set<Resource> getResources() {
+        return resources;
+    }
+
     public List<FileScope> getFileScopes() {
         return fileScopes;
     }
@@ -238,13 +243,15 @@ public class RootScope extends FileScope {
     }
 
     public List<Resource> findResourcesIn(Set<String> diffFiles) {
-        Stream<Resource> stream = Stream.concat(Stream.of(this), getFileScopes().stream())
+        Stream<Resource> stream = Stream.of(this)
             .map(Map::entrySet)
             .flatMap(Collection::stream)
             .filter(e -> e.getValue() instanceof Resource)
             .filter(e -> ((Resource) e.getValue()).primaryKey().equals(e.getKey()))
             .map(Entry::getValue)
             .map(Resource.class::cast);
+
+        stream = Stream.concat(stream, getResources().stream());
 
         if (diffFiles != null && !diffFiles.isEmpty()) {
             stream = stream.filter(r -> diffFiles.contains(DiffableInternals.getScope(r).getFileScope().getFile()));

--- a/core/src/main/java/gyro/core/scope/RootScope.java
+++ b/core/src/main/java/gyro/core/scope/RootScope.java
@@ -20,6 +20,7 @@ import java.io.IOException;
 import java.lang.reflect.Type;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.LinkedHashMap;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
@@ -102,7 +103,7 @@ public class RootScope extends FileScope {
     private final FileBackend backend;
     private final RootScope current;
     private final Set<String> loadFiles;
-    private final Set<Resource> resources = new LinkedHashSet<>();
+    private final Map<String, Resource> resources = new LinkedHashMap<>();
     private final List<FileScope> fileScopes = new ArrayList<>();
 
     public RootScope(String file, FileBackend backend, RootScope current, Set<String> loadFiles) {
@@ -196,7 +197,7 @@ public class RootScope extends FileScope {
         return loadFiles;
     }
 
-    public Set<Resource> getResources() {
+    public Map<String, Resource> getResources() {
         return resources;
     }
 
@@ -238,20 +239,18 @@ public class RootScope extends FileScope {
         return converter.convert(returnType, object);
     }
 
-    public List<Resource> findResources() {
-        return findResourcesIn(null);
+    public List<Resource> findSortedResources() {
+        return findSortedResourcesIn(null);
     }
 
-    public List<Resource> findResourcesIn(Set<String> diffFiles) {
-        Stream<Resource> stream = Stream.of(this)
+    public List<Resource> findSortedResourcesIn(Set<String> diffFiles) {
+        Stream<Resource> stream = Stream.concat(Stream.of(this), Stream.of(getResources()))
             .map(Map::entrySet)
             .flatMap(Collection::stream)
             .filter(e -> e.getValue() instanceof Resource)
             .filter(e -> ((Resource) e.getValue()).primaryKey().equals(e.getKey()))
             .map(Entry::getValue)
             .map(Resource.class::cast);
-
-        stream = Stream.concat(stream, getResources().stream());
 
         if (diffFiles != null && !diffFiles.isEmpty()) {
             stream = stream.filter(r -> diffFiles.contains(DiffableInternals.getScope(r).getFileScope().getFile()));
@@ -261,7 +260,7 @@ public class RootScope extends FileScope {
     }
 
     public <T extends Resource> Stream<T> findResourcesByClass(Class<T> resourceClass) {
-        return findResources()
+        return findSortedResources()
             .stream()
             .filter(resourceClass::isInstance)
             .map(resourceClass::cast);
@@ -386,7 +385,7 @@ public class RootScope extends FileScope {
             throw new GyroException(sb.toString());
         }
 
-        List<ValidationError> errors = findResources().stream()
+        List<ValidationError> errors = findSortedResources().stream()
             .map(r -> DiffableType.getInstance(r).validate(r))
             .flatMap(List::stream)
             .collect(Collectors.toList());

--- a/core/src/main/java/gyro/core/virtual/VirtualResourceVisitor.java
+++ b/core/src/main/java/gyro/core/virtual/VirtualResourceVisitor.java
@@ -90,7 +90,7 @@ public class VirtualResourceVisitor extends ResourceVisitor {
 
         String prefix = name + "/";
 
-        for (Resource resource : virtualRoot.findResources()) {
+        for (Resource resource : virtualRoot.findSortedResources()) {
             DiffableInternals.setName(resource, prefix + DiffableInternals.getName(resource));
             file.put(resource.primaryKey(), resource);
         }

--- a/core/src/main/java/gyro/core/workflow/Stage.java
+++ b/core/src/main/java/gyro/core/workflow/Stage.java
@@ -109,8 +109,8 @@ public class Stage {
         apply(ui, state, currentResource, pendingResource, pendingRootScope);
 
         Diff diff = new Diff(
-            currentRootScope.findResourcesIn(currentRootScope.getLoadFiles()),
-            pendingRootScope.findResourcesIn(pendingRootScope.getLoadFiles()));
+            currentRootScope.findSortedResourcesIn(currentRootScope.getLoadFiles()),
+            pendingRootScope.findSortedResourcesIn(pendingRootScope.getLoadFiles()));
 
         diff.diff();
 

